torch/utils/flop_counter.py

```diff
@@ -0,0 +1,306 @@
+    # Return the value and the suffix as a string
+    return value + suffixes[index]
+
+class FlopCounterMode(TorchDispatchMode):
+    """
+    ``FlopCounterMode`` is a context manager that counts the number of
+    flops within its context. It does this using a ``TorchDispatchMode``.
+
+    It also supports hierarchical output by passing a module to FlopCounterMode on construction.
+
+    Example usage
+
+    .. code-block:: python
+
+        mod = ...
+        flop_counter = FlopCounterMode(mod)
+        with flop_counter:
+            mod.sum().backward()
+
+    """
+    def __init__(self, mod: torch.nn.Module = None, depth: int = 2, display: bool = True, custom_mapping: Dict[Any, Any] = None):
+        self.flop_counts: Dict[Any, Any] = defaultdict(lambda: defaultdict(int))
+        self.depth = depth
+        self.parents = ["Global"]
+        self.display = display
+        if custom_mapping is None:
+            custom_mapping = {}
+        if mod is not None:
+            for name, module in dict(mod.named_modules()).items():
+                module.register_forward_pre_hook(self._enter_module(name))
+                module.register_forward_hook(self._exit_module(name))
+        self.flop_mapping = {**flop_mapping, **custom_mapping}
+
+    def _enter_module(self, name):
+        def f(module, inputs):
+            self.parents.append(name)
+            inputs = normalize_tuple(inputs)
+            out = self._create_backwards_pop(name)(*inputs)
+            return out
+
+        return f
+
+    def _exit_module(self, name):
+        def f(module, inputs, outputs):
+            assert(self.parents[-1] == name)
+            self.parents.pop()
+            outputs = normalize_tuple(outputs)
+            return self._create_backwards_push(name)(*outputs)
+        return f
+
+    def _create_backwards_push(self, name):
+        class PushState(torch.autograd.Function):
+            @staticmethod
+            def forward(ctx, *args):
+                args = tree_map(lambda x: x.clone() if isinstance(x, torch.Tensor) else x, args)
+                if len(args) == 1:
+                    return args[0]
+                return args
+
+            @staticmethod
+            def backward(ctx, *grad_outs):
+                self.parents.append(name)
+                return grad_outs
+
+        return PushState.apply
+
+    def _create_backwards_pop(self, name):
+        class PopState(torch.autograd.Function):
+            @staticmethod
+            def forward(ctx, *args):
+                args = tree_map(lambda x: x.clone() if isinstance(x, torch.Tensor) else x, args)
+                if len(args) == 1:
+                    return args[0]
+                return args
+
+            @staticmethod
+            def backward(ctx, *grad_outs):
+                assert(self.parents[-1] == name)
+                self.parents.pop()
+                return grad_outs
+
+        return PopState.apply
+
+    def register_hierarchy(self, f, name):
```

Comment:
What is this doing?

Reply:
Hmm... it's just providing a way for users to register a custom level in the hierarchy. Perhaps it's unneeded tbh - might remove it.